Code/GraphMol/MarvinParse/MarvinDefs.cpp

```diff
@@ -0,0 +1,2978 @@
+  while (actualParent->role() == "MultipleSgroup") {
+    actualParent = actualParent->parent;
+  }
+
+  for (MarvinBond *bondPtr : actualParent->bonds) {
+    bool atom1InSet = boost::algorithm::contains(
+        this->atoms, std::vector<std::string>{bondPtr->atomRefs2[0]},
+        atomRefInAtoms);
+    bool atom2InSet = boost::algorithm::contains(
+        this->atoms, std::vector<std::string>{bondPtr->atomRefs2[1]},
+        atomRefInAtoms);
+    if (atom1InSet && atom2InSet) {
+      bondsInGroup.push_back(bondPtr);
+    } else if (atom1InSet) {
+      originalConnectorAtoms.push_back(bondPtr->atomRefs2[0]);
+      this->bondsToAtomsNotInExpandedGroup.push_back(bondPtr);
+      outsideConnectorAtom =
+          bondPtr
+              ->atomRefs2[1];  // last one set wins - when done should be the
+                               // outside atom for the 2nd bond to the outside
+    } else if (atom2InSet) {
+      originalConnectorAtoms.push_back(bondPtr->atomRefs2[1]);
+      this->bondsToAtomsNotInExpandedGroup.push_back(bondPtr);
+      outsideConnectorAtom =
+          bondPtr
+              ->atomRefs2[0];  // last one set wins - when done should be the
+                               // outside atom for the 2nd bond to the outside
+    }
+  }
+
+  // should be two bonds to the outside (or zero)
+  if (this->bondsToAtomsNotInExpandedGroup.size() != 2 &&
+      this->bondsToAtomsNotInExpandedGroup.size() != 0) {
+    throw FileParseException(
+        "Error - there must be zero or two bonds from the group to atoms not in the group");
+  }
+
+  // find the hightest id of an atom in the Group - we will insert replicated
+  // atoms after that one
+
+  int highestIndex = 0;
+  for (MarvinAtom *atomPtr : this->atoms) {
+    int thisIndex = actualParent->getAtomIndex(atomPtr->id);
+    if (thisIndex > highestIndex) {
+      highestIndex = thisIndex;
+    }
+  }
+  MarvinAtom *lastAtomInGroupPtr = actualParent->atoms[highestIndex];
+
+  // ready to make copies of the group
+
+  int copyCount = std::stoi(this->title);
+
+  std::vector<MarvinMolBase *> sgroupsToAdd;
+  if (this->bondsToAtomsNotInExpandedGroup.size() == 2) {
+    connectorAtoms[1] =
+        originalConnectorAtoms[1];  // this one does NOT have an _# on it
+  }
+
+  for (int copyIndex = 2; copyIndex <= copyCount;
+       ++copyIndex)  // start at 2, we already have the first copy
+  {
+    std::string idAppendage = "_" + this->id + ":" + std::to_string(copyIndex);
+    for (auto parentAtomPtr : this->parentAtoms) {
+      auto copyAtom =
+          new MarvinAtom(*parentAtomPtr, parentAtomPtr->id + idAppendage);
+      if (parentAtomPtr->sgroupRef != "" && !copyAtom->sGroupRefIsSuperatom) {
+        copyAtom->sgroupRef = parentAtomPtr->sgroupRef + idAppendage;
+      }
+
+      // copy atoms to all parents up to the actual parent (multipleSgrups in
+      // multipleSgroups)
+
+      for (auto thisParent = this->parent;; thisParent = thisParent->parent) {
```

Comment:
It seems like it would be a lot safer here (and in all the other places where you use this pattern for iterating) to use the second part of the `for` loop to make sure that `thisParent` isn't null.
```suggestion
      for (auto thisParent = this->parent;thisParent != nullptr; thisParent = thisParent->parent) {
```
This little extra bit of defensive programming doesn't make the code slower or more complicated, but it could prevent a seg fault if something weird happens

Reply:
this needs at least a reply

Reply:
I certainly do NOT want this loop to exit and not throw and error if the pointer is nullptr.  if something weird happens, the next line would throw an error.  I could put a TEST_ASSERT as the first line in the loop, but  it is going to throw an error anyway.

Reply:
The next line will seg fault when it tries to dereference the null ptr.
If you want to throw an error instead, then please either use `TEST_ASSERT` (it's non-recoverable and likely a programming error) or throw a `FileParseException`

Reply:
OK done